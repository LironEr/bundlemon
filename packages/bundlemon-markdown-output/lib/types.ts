export interface GenerateReportMarkdownOptions {
  showLinkToReport?: boolean;
}
