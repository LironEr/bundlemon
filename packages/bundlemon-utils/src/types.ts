/* istanbul ignore file */

import { Compression, DiffChange, FailReason, Status } from './consts';

export interface FileDetails {
  /**
   * @minLength 1
   * @maxLength 50
   */
  friendlyName?: string;
  pattern: string;
  path: string;
  size: number;
  compression: Compression;
  maxSize?: number;
  maxPercentIncrease?: number;
}

export interface DiffFromBase {
  bytes: number;
  percent: number;
  change: DiffChange;
}

export type FileStatusObject =
  | { status: Status.Pass; failReasons?: undefined }
  | { status: Status.Fail; failReasons: FailReason[] };

export type FileDetailsDiff = FileDetails & { diff: DiffFromBase } & FileStatusObject;

export interface CommitRecordPayload {
  /**
   * @minLength 1
   * @maxLength 100
   * @pattern ^[A-Za-z0-9_\-. ]*$
   */
  subProject?: string;
  files: FileDetails[];
  groups: FileDetails[];
  /**
   * @minLength 1
   * @maxLength 100
   */
  branch: string;
  /**
   * @minLength 1
   * @maxLength 100
   * @pattern ^[A-Za-z0-9]*$
   */
  commitSha: string;
  /**
   * @minLength 1
   * @maxLength 100
   */
  baseBranch?: string;
  /**
   * @minLength 1
   * @maxLength 10
   * @pattern ^[0-9]*$
   */
  prNumber?: string;
}

export interface CommitRecord extends CommitRecordPayload {
  id: string;
  projectId: string;
  creationDate: string;
}

export interface DiffStats {
  currBranchSize: number;
  baseBranchSize: number;
  diff: {
    bytes: number;
    percent: number;
  };
}

export interface BaseCommitRecordResponse {
  record: CommitRecord;
  baseRecord?: CommitRecord;
}

export interface CreateCommitRecordResponse extends BaseCommitRecordResponse {
  linkToReport: string;
}

export interface ReportMetadata {
  subProject?: string;
  linkToReport?: string;
  record?: CommitRecord;
  baseRecord?: CommitRecord;
}

export interface DiffReportInput {
  files: FileDetails[];
  groups: FileDetails[];
}

export interface DiffReport {
  files: FileDetailsDiff[];
  stats: DiffStats;
  groups: FileDetailsDiff[];
  status: Status;
}

export interface Report extends DiffReport {
  metadata: ReportMetadata;
}

export interface CreateProjectResponse {
  projectId: string;
  apiKey: string;
}

export type OutputResult = 'success' | 'failure' | 'skipped';

export interface OutputResponse {
  result: OutputResult;
  message: string;
  metadata?: Record<string, unknown>;
}
export type GithubOutputTypes = 'checkRun' | 'commitStatus' | 'prComment';
export type GithubOutputResponse = Partial<Record<GithubOutputTypes, OutputResponse>>;
