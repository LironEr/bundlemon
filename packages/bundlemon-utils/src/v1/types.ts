/* istanbul ignore file */

export interface FileDetails {
  pattern: string;
  path: string;
  size: number;
  maxSize?: number;
  maxPercentIncrease?: number;
}

export interface ReportPayload {
  files: FileDetails[];
  defaultCompression: 'gzip' | 'none';
  branch: string;
  commitSha: string;
  baseBranch?: string;
  prNumber?: string;
}

export interface Report extends Omit<ReportPayload, 'defaultCompression'> {
  id: string;
  projectId: string;
  creationDate: string;
}

export interface BaseReportResponse {
  report: Report;
  baseReport?: Report;
}

export interface CreateReportResponse extends BaseReportResponse {
  linkToReport: string;
}

export interface CreateProjectResponse {
  projectId: string;
  apiKey: string;
}
