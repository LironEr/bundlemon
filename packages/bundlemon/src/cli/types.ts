import type { Compression } from 'bundlemon-utils';

export interface CliOptions {
  config?: string;
  subProject?: string;
  defaultCompression?: Compression;
}
