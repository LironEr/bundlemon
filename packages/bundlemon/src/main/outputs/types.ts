import type { Report } from 'bundlemon-utils';
import type { NormalizedConfig } from '../types';

export interface OutputCreateParams {
  config: NormalizedConfig;
  options: unknown;
}

export interface OutputInstance {
  generate(report: Report): Promise<void> | void;
}

export interface Output {
  name: string;
  create: (params: OutputCreateParams) => OutputInstance | undefined | Promise<OutputInstance | undefined>;
}
