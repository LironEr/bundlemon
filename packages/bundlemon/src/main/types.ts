import type { Compression, ProjectProvider } from 'bundlemon-utils';
import type { CreateCommitRecordAuthType } from '../common/consts';

export interface FileConfig {
  friendlyName?: string;
  path: string;
  compression?: Compression;
  maxSize?: string;
  maxPercentIncrease?: number;
}

export interface NormalizedFileConfig extends Omit<FileConfig, 'maxSize' | 'compression'> {
  compression: Compression;
  maxSize?: number;
}

export interface Config {
  subProject?: string;
  baseDir?: string;
  files: FileConfig[];
  groups?: FileConfig[];
  verbose?: boolean;
  defaultCompression?: Compression;
  reportOutput?: (string | [string, unknown])[];
}

export interface BaseNormalizedConfig extends Omit<Required<Config>, 'files' | 'groups' | 'subProject'> {
  subProject?: string;
  files: NormalizedFileConfig[];
  groups: NormalizedFileConfig[];
  remote: boolean;
}

export interface NormalizedConfigRemoteOn extends BaseNormalizedConfig {
  remote: true;
  projectId: string;
  gitVars: GitVars;
  getCreateCommitRecordAuthParams: () => CreateCommitRecordAuthParams;
}

export interface NormalizedConfigRemoteOff extends BaseNormalizedConfig {
  remote: false;
}

export type NormalizedConfig = NormalizedConfigRemoteOn | NormalizedConfigRemoteOff;

export interface MatchFile {
  fullPath: string;
  prettyPath: string;
}

export interface GitVars {
  branch: string;
  commitSha: string;
  baseBranch?: string;
  prNumber?: string;
}

export type CreateCommitRecordProjectApiKeyAuthQuery = {
  authType: CreateCommitRecordAuthType.ProjectApiKey;
  token: string;
};

export type CreateCommitRecordGithubActionsAuthQuery = {
  authType: CreateCommitRecordAuthType.GithubActions;
  runId: string;
};

export type CreateCommitRecordAuthParams =
  | CreateCommitRecordProjectApiKeyAuthQuery
  | CreateCommitRecordGithubActionsAuthQuery;

export interface GitDetails {
  provider: ProjectProvider;
  owner: string;
  repo: string;
}
