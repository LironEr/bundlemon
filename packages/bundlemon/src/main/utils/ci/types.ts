export interface CIEnvVars {
  owner?: string;
  repo?: string;
  branch?: string;
  commitSha?: string;
  targetBranch?: string;
  prNumber?: string;
}

export interface Provider {
  isItMe: boolean;
  getVars: () => CIEnvVars;
}
