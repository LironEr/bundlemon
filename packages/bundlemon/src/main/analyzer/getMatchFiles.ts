import * as path from 'path';
import * as fs from 'fs';
import * as micromatch from 'micromatch';

import type { MatchFile, NormalizedFileConfig } from '../types';

export async function getAllPaths(dirPath: string): Promise<string[]> {
  const entries = await fs.promises.readdir(dirPath, { withFileTypes: true });

  const files = entries.filter((file) => !file.isDirectory()).map((file) => path.join(dirPath, file.name));
  const folders = entries.filter((folder) => folder.isDirectory());

  for (const folder of folders) files.push(...(await getAllPaths(path.join(dirPath, folder.name))));

  return files;
}

export function createPrettyPath(filePath: string, globPattern: string): string {
  let prettyPath = filePath;
  const re = micromatch.makeRe(globPattern);
  const groups = re.exec(filePath)?.groups || {};

  Object.entries(groups).forEach(([key, value]) => {
    if (key.startsWith('hash')) {
      prettyPath = prettyPath.replace(value, '(hash)');
    }
  });

  return prettyPath;
}

export function getRegexHash(index: number): string {
  return `(?<hash${index}>[a-zA-Z0-9]+)`;
}

export async function getMatchFiles(baseDir: string, filesConfig: NormalizedFileConfig[]): Promise<MatchFile[]> {
  const transformedFilesConfig = filesConfig.map((f) => {
    const { path: filePattern, ...rest } = f;
    let index = 0;

    return { ...rest, path: filePattern.replace(/<hash>/g, () => getRegexHash(index++)) };
  });

  const allFiles = await getAllPaths(baseDir);

  const matchFiles: MatchFile[] = [];

  for (const fullPath of allFiles) {
    const relativePath = path.relative(baseDir, fullPath);

    for (const fileConfig of transformedFilesConfig) {
      const { path: pattern, ...restConfig } = fileConfig;

      if (micromatch.isMatch(relativePath, pattern)) {
        const prettyPath = createPrettyPath(relativePath, pattern);

        matchFiles.push({ ...restConfig, fullPath, prettyPath });

        break;
      }
    }
  }

  return matchFiles;
}
