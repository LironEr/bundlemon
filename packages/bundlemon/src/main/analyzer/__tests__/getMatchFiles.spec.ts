import * as fs from 'fs';
import * as path from 'path';
import { getAllPaths, createPrettyPath, getRegexHash, getMatchFiles } from '../getMatchFiles';
import { NormalizedFileConfig, MatchFile } from '../../types';

function matchFileComparer(a: MatchFile, b: MatchFile) {
  return a.fullPath.localeCompare(b.fullPath);
}

describe('getFiles', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('getAllPaths', () => {
    it('recursive directories', async () => {
      const paths = await getAllPaths(__dirname + '/fixtures/getAllPaths/1');

      const fileNames = paths.map((fullPath) => path.basename(fullPath));

      expect(fileNames.sort()).toEqual(['a.html', 'a.js', 'ac.css', 'g.aa.bbb'].sort());
    });

    it('empty directory', async () => {
      const dirPath = __dirname + '/fixtures/getAllPaths/2';
      if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath);
      }

      const paths = await getAllPaths(dirPath);

      expect(paths).toMatchSnapshot();
    });
  });

  describe('createPrettyPath', () => {
    it('simple file name', async () => {
      expect(createPrettyPath('index.html', 'index.html')).toEqual('index.html');
      expect(createPrettyPath('a/index.html', 'test')).toEqual('a/index.html');
    });

    it('not match glob pettern -> doesnt change path', async () => {
      expect(createPrettyPath('a/index.html', 'test')).toEqual('a/index.html');
    });

    it('glob star', async () => {
      expect(createPrettyPath('index.html', '**/*')).toEqual('index.html');
      expect(createPrettyPath('a/b/c.ts', '**/*')).toEqual('a/b/c.ts');
    });

    it('transform hash', async () => {
      expect(createPrettyPath('main.jhag2djh.css', '**/*.(?<hash0>[a-zA-Z0-9]+).css')).toEqual('main.(hash).css');
      expect(createPrettyPath('a/b/main.hasdhj.css', '**/*.(?<hash0>[a-zA-Z0-9]+).css')).toEqual('a/b/main.(hash).css');
    });

    it('transform multiple hashes', async () => {
      expect(
        createPrettyPath(
          'jhasdhj/a/test.hjas2djh.chunk.css',
          '(?<hash0>[a-zA-Z0-9]+)/**/*.(?<hash1>[a-zA-Z0-9]+).chunk.css'
        )
      ).toEqual('(hash)/a/test.(hash).chunk.css');

      expect(
        createPrettyPath(
          'jhasdhj/test.252343.chunk.css',
          '(?<hash0>[a-zA-Z0-9]+)/**/*.(?<hash1>[a-zA-Z0-9]+).chunk.css'
        )
      ).toEqual('(hash)/test.(hash).chunk.css');

      expect(createPrettyPath('jh2asjhd.252s343.css', '(?<hash0>[a-zA-Z0-9]+).(?<hash1>[a-zA-Z0-9]+).css')).toEqual(
        '(hash).(hash).css'
      );
    });

    it('transform only hashes', async () => {
      expect(
        createPrettyPath(
          'jhasdhj/a/test.aaaaa.chunk.css',
          '(?<hash0>[a-zA-Z0-9]+)/**/*.(?<group0>[a-zA-Z0-9]+).(?<hash1>[a-zA-Z0-9]+).css'
        )
      ).toEqual('(hash)/a/test.aaaaa.(hash).css');
    });
  });

  it('getRegexHash', async () => {
    expect(getRegexHash(0)).toMatchSnapshot();
    expect(getRegexHash(0)).toMatchSnapshot();
  });

  describe('getMatchFiles', () => {
    it('recursive directories', async () => {
      const fixturePath = __dirname + '/fixtures/getMatchFiles';

      const config: NormalizedFileConfig[] = [
        {
          path: 'index.html',
          maxSize: 15000,
        },
        {
          path: '**/*.<hash>.chunk.js',
          maxSize: 10000,
        },
        {
          path: '**/main.<hash>.js',
          maxSize: 100000,
        },
        {
          path: '**/*.<hash>.js',
          maxSize: 50000,
        },
        {
          path: '**/*.js',
          maxSize: 5000,
        },
      ];

      const expectedMatchFiles: MatchFile[] = [
        {
          maxSize: 15000,
          fullPath: path.join(fixturePath, '/index.html'),
          prettyPath: 'index.html',
        },
        {
          maxSize: 5000,
          fullPath: path.join(fixturePath, '/service.js'),
          prettyPath: 'service.js',
        },
        {
          maxSize: 50000,
          fullPath: path.join(fixturePath, '/static/js/about.hjasj2u.js'),
          prettyPath: 'static/js/about.(hash).js',
        },
        {
          maxSize: 50000,
          fullPath: path.join(fixturePath, '/static/js/login.a2j21i.js'),
          prettyPath: 'static/js/login.(hash).js',
        },
        {
          maxSize: 100000,
          fullPath: path.join(fixturePath, '/static/js/main.jh2j2ks.js'),
          prettyPath: 'static/js/main.(hash).js',
        },
        {
          maxSize: 5000,
          fullPath: path.join(fixturePath, '/static/js/other.js'),
          prettyPath: 'static/js/other.js',
        },
        {
          maxSize: 10000,
          fullPath: path.join(fixturePath, '/static/js/test.jks22892s.chunk.js'),
          prettyPath: 'static/js/test.(hash).chunk.js',
        },
        {
          maxSize: 10000,
          fullPath: path.join(fixturePath, '/static/js/test2.js2k2kxj.chunk.js'),
          prettyPath: 'static/js/test2.(hash).chunk.js',
        },
      ].sort(matchFileComparer);

      const matchFiles = await getMatchFiles(fixturePath, config);

      const sortedMatchFiles = matchFiles.sort(matchFileComparer);

      expect(sortedMatchFiles).toEqual(expectedMatchFiles);
    });
  });
});
