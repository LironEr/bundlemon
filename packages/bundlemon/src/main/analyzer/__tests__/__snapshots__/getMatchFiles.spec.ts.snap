// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFiles getAllPaths empty directory 1`] = `Array []`;

exports[`getFiles getAllPaths recursive directories 1`] = `
Array [
  "/Users/lironer/Documents/projects/bundlemon/bundlemon/packages/bundlemon/src/main/analyzer/__tests__/assets/getAllPaths/1/a.html",
  "/Users/lironer/Documents/projects/bundlemon/bundlemon/packages/bundlemon/src/main/analyzer/__tests__/assets/getAllPaths/1/a.js",
  "/Users/lironer/Documents/projects/bundlemon/bundlemon/packages/bundlemon/src/main/analyzer/__tests__/assets/getAllPaths/1/s/g.aa.bbb",
  "/Users/lironer/Documents/projects/bundlemon/bundlemon/packages/bundlemon/src/main/analyzer/__tests__/assets/getAllPaths/1/s/c/ac.css",
]
`;

exports[`getFiles getRegexHash 1`] = `"(?<hash0>[a-zA-Z0-9]+)"`;

exports[`getFiles getRegexHash 2`] = `"(?<hash0>[a-zA-Z0-9]+)"`;
